### DISCUSSION TITLE
Parallel.ForEach - Access Violation: 'Attempted to read or write protected memory'

### DISCUSSION THREAD

**Discussion by TJPovey** [MaxRev-Dev/gdal.netcore#92] [Q&A] [answered]
I have millions of points which I have been transforming using this library, which works great.

Because there are so many points to transform, I would like to use parallelism, using ```Parallel.ForEach```.
This results in **Access Violation: 'Attempted to read or write protected memory'**. Since the coordinate transformation is created outside of the loop. Creating the coordinate transformation in the loop is too much of an overhead and takes too much time.

I can see that the CoordinateTransformation class takes an optional options parameter: CoordinateTransformationOptions, which takes some pointers, but I'm not sure how to get this working/if it is even a fix to the above.

Example code:

```csharp
var transformation = new CoordinateTransformation(fromCS, toCS);

Parallel.ForEach(points, point => {
    var transformedPoint = new double[3];
    transformation.TransformPoint(transformedPoint, point.x, point.y, point.z);
});
```

I know you can pass in an array of points, which will most likely be more efficient, but what I'm doing involves several other calculations on each point, e.g. multiplying it by a matrix, creating a new vertex, and assigning it to a Mesh.

Is it possible to achieve parallelism using Gdal?

Thanks!

**Comment by MaxRev-Dev** [comment, marked as answer]
That API only supports one point for a method call. Also, there can be issues with P/Invoke thread safety. In my opinion, it's better to use [Gdal.Warp](
https://gdal.org/programs/gdalwarp.html) or [GDALVectorTranslate](https://gdal.org/api/gdal_utils.html#_CPPv419GDALVectorTranslatePKc12GDALDatasetHiP12GDALDatasetHPK26GDALVectorTranslateOptionsPi) for this purpose. You can also consider using a native library [ProjNet](https://github.com/NetTopologySuite/ProjNet4GeoAPI).

**Reply by donnyv** [reply]
Is there any sample code to project Gdal geometries using ProjNet? It looks like it only takes an array of points. That can get pretty wild if you dont know the geometry you are projecting. You would have to write the code to rebuild every geometry type.
